depscan/lib/utils.py

```diff
@@ -95,90 +70,42 @@ def is_exe(src):
-        return ["go", "binary"]
+    if os.path.isfile(src_dir):
+        return detect_project_type_from_file(src_dir)
+    
     project_types = []
-    if find_python_reqfiles(src_dir) or find_files(src_dir, ".py", quick=True):
-        project_types.append("python")
-    if find_files(src_dir, "pom.xml", quick=True) or find_files(
-        src_dir, ".gradle", quick=True
-    ):
-        project_types.append("java")
-    if find_files(src_dir, ".gradle.kts", quick=True):
-        project_types.append("kotlin")
-    if find_files(src_dir, "build.sbt", quick=True):
-        project_types.append("scala")
-    if (
-        find_files(src_dir, "package.json", quick=True)
-        or find_files(src_dir, "yarn.lock", quick=True)
-        or find_files(src_dir, "rush.json", quick=True)
-    ):
-        project_types.append("nodejs")
-    if find_files(src_dir, "go.sum", quick=True) or find_files(
-        src_dir, "Gopkg.lock", quick=True
-    ):
-        project_types.append("go")
-    if find_files(src_dir, "Cargo.lock", quick=True):
-        project_types.append("rust")
-    if find_files(src_dir, "composer.json", quick=True):
-        project_types.append("php")
-    if find_files(src_dir, ".csproj", quick=True):
-        project_types.append("dotnet")
-    if find_files(src_dir, "Gemfile", quick=True) or find_files(
-        src_dir, "Gemfile.lock", quick=True
-    ):
-        project_types.append("ruby")
-    if find_files(src_dir, "deps.edn", quick=True) or find_files(
-        src_dir, "project.clj", quick=True
-    ):
-        project_types.append("clojure")
-    if find_files(src_dir, "conan.lock", quick=True) or find_files(
-        src_dir, "conanfile.txt", quick=True
-    ):
-        project_types.append("cpp")
-    if find_files(src_dir, "pubspec.lock", quick=True) or find_files(
-        src_dir, "pubspec.yaml", quick=True
-    ):
-        project_types.append("dart")
-    if find_files(src_dir, "cabal.project.freeze", quick=True):
-        project_types.append("haskell")
-    if find_files(src_dir, "mix.lock", quick=True):
-        project_types.append("elixir")
-    if find_files(
-        os.path.join(src_dir, ".github", "workflows"),
-        ".yml",
-        quick=True,
-        filter=False,
-    ):
-        project_types.append("github")
-    # jars
-    if "java" not in project_types and find_files(src_dir, ".jar", quick=True):
-        project_types.append("jar")
-    # Jenkins plugins or plain old jars
-    if "java" not in project_types and find_files(src_dir, ".hpi", quick=True):
-        project_types.append("jenkins")
-    if find_files(src_dir, ".yml", quick=True) or find_files(
-        src_dir, ".yaml", quick=True
-    ):
-        project_types.append("yaml-manifest")
-    return project_types
+    for root, dirs, files in os.walk(src_dir):
+        filter_ignored_dirs(dirs)
+        for file in files:
+            project_types.extend(detect_project_type_from_file(os.path.abspath(os.path.join(root, file))))
```

Comment:
Can we add a quick mode so we don't check every file?

Reply:
@saketjajoo Is this feasible?

Reply:
Theoretically, yes this is feasible. I can just tweak the original implementation to include everything, but the code readability would get affected (as I would have to rewrite the same logic twice - one for filename as input and one for directory as input). However, in this approach, the code is a bit more readable and also more accurate but a bit slow as compared to the original one. Hence there's a tradeoff between (code readability + correctness) vs performance.

Reply:
@saketjajoo @cerrussell Shall we explore an alternative where depscan doesn't attempt any detection at all but simply invokes cdxgen with either `-t universal` or without any type when `--no-universal` is passed? This way we reuse the type detection performed by cdxgen.

Reply:
@prabhu Does that mean we couldn't use --bom?

Reply:
When invoked with --bom, can we parse the bom document to collect the types?